django/core/validators.py

```diff
@@ -219,8 +218,16 @@ class EmailValidator:
         re.IGNORECASE,
     )
     domain_regex = _lazy_re_compile(
-        # max length for domain name labels is 63 characters per RFC 1034
-        r"((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))\Z",
+        # Standard ASCII domain pattern
+        r"(?:"
+        r"((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))"
+        r"|"
+        # IDN pattern
+        r"(?:" + DomainNameValidator.hostname_re + DomainNameValidator.domain_re + r"\."
+        r"(?!-)"
+        r"(?:[a-z" + DomainNameValidator.ul + "-]{2,63}|xn--[a-z0-9]{1,59})"
+        r"(?<!-))"
+        r")\Z",
```

Comment:
The "IDN" pattern *also* already allows standard ASCII domains, so shouldn't be a need to break out ASCII separately: 

```suggestion
        r"^" + host_re + r"\Z",
```

Where `host_re = URLValidator.host_re`. (I'd suggest adding a line that makes that "borrowing" explicit just above the domain_regex, similar to how borrowing is done [in URLValidator](https://github.com/django/django/blob/f05edb2b43c347d4929efd52c8e2b4e08839f542/django/core/validators.py#L138-L140).)